Remove viewmodel dependencies

arcanerogue opened the pull request:
Prior to this PR, my ViewModels contained PhunApp and EventRepository dependencies. Greg recommended that I use the Dependency Injection pattern to pass in any needed dependency references as parameters in my ViewModel constructors.

- I’ve refactored my ViewModels, EventViewModel and EventDetailViewModel, to extend the ViewModel class rather than extending the AndroidViewModel class.  Previously, I extended the AndroidViewModel class so that I could pass in a reference to the PhunApp instance to expose the reference to the EventRepository instance.  

  By extending the ViewModel class and passing in the EventRepository reference as a parameter in the ViewModel constructors, I am removing my previous PhunApp and EventRepository dependencies.

- I’ve created a ViewModelFactory class to handle the creation of my ViewModel class and passing in the EventRepository dependency in the ViewModel constructors.  The ViewModelFactory class extends the ViewModelProvider.NewInstanceFactory class, to allow creation of different types of ViewModels (EventViewModel and EventDetailViewModel) which will be decided at runtime.  If a ViewModel class is requested that does not match these two types, an IllegalArgumentException will be thrown.  

- The creation of a EventViewModel requested by the MainActivity should be as follows:
<img width="800" alt="Screen Shot 2019-05-23 at 12 32 54 AM" src="https://user-images.githubusercontent.com/13700908/58234046-76e28f80-7cf2-11e9-8283-48c156182ac6.png">
 
- I've renamed the com.glopez.phunapp.ui package name to com.glopez.phunapp.view

- I've renamed the com.glopez.phunapp.date package name to com.glopez.phunapp.model

grasmussen-phunware commented on app/src/main/res/layout-land/content_main.xml:
```diff
@@ -0,0 +1,27 @@
+            android:id="@+id/feed_list"
+            android:layout_width="0dp"
+            android:layout_height="0dp"
+            android:layout_marginTop="8dp"
```
Hi @arcanerogue.  You don't need to worry about it now, but to let you know, in the long run, the common dimensions, such as 8dp, should be externalized to the dimensions files (dimens.xml).

arcanerogue replied:
Thanks for the heads up @grasmussen-phunware!

grasmussen-phunware commented on app/src/main/res/layout-land/content_main.xml:
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="utf-8"?>
+<android.support.constraint.ConstraintLayout
```
👍

grasmussen-phunware commented on app/src/main/java/com/glopez/phunapp/view/viewmodels/EventViewModel.kt:
```diff
@@ -0,0 +1,23 @@
+ * The ViewModel for the MainActivity which displays the list of Event objects.
+ * @param[eventRepo] The application's repository instance.
+ */
+class EventViewModel(private val eventRepo: EventRepository) : ViewModel() {
```
👍

grasmussen-phunware commented on app/src/main/java/com/glopez/phunapp/view/viewmodels/EventViewModel.kt:
```diff
@@ -0,0 +1,23 @@
+ */
+class EventViewModel(private val eventRepo: EventRepository) : ViewModel() {
+    val events: LiveData<List<Event>>
+        get() = getEventsList()
```
Hi @arcanerogue .  Since there is already a getter function, named getEventsList, consider removing this get function and making the `events` val private.  I recommend figuring out how to combine them, or remove this one.

arcanerogue replied:
👍

grasmussen-phunware commented on app/src/main/java/com/glopez/phunapp/view/viewmodels/EventDetailViewModel.kt:
```diff
@@ -0,0 +1,22 @@
+     * @param[id] The value of the id property of the requested Event.
+     * @return The requested Event object.
+     */
+    fun getEvent(id: Int): LiveData<Event> {
```
Hi @arcanerogue .  I recommend thinking about what could happen if the id is invalid and an event isn't found.

arcanerogue replied:
@grasmussen-phunware, I've been thinking about several approaches to this and would like to discuss with you when you have some time.

grasmussen-phunware commented on app/src/main/java/com/glopez/phunapp/view/activities/MainActivity.kt:
```diff
@@ -33,12 +35,14 @@ class MainActivity : AppCompatActivity() {
-                // Display toast when there was no data retrieved from the database and
+                // Display toast when there was no model retrieved from the database and
                 // a network connection is unavailable.
-//                if (events.isEmpty() && !isNetworkAvailable(this)) {
```
👍 I always promote removing commented out code.  It helps keep the codebase clean.

arcanerogue replied:
Ah I missed that one!  Good catch!

grasmussen-phunware replied:
No prob. 😄

grasmussen-phunware commented on app/src/main/java/com/glopez/phunapp/ViewModelFactory.kt:
```diff
@@ -0,0 +1,53 @@
+         * @return The ViewModelFactory instance.
+         */
+        fun getInstance(application: PhunApp): ViewModelFactory {
+            synchronized(this) {
```
👏

grasmussen-phunware commented on app/build.gradle:
```diff
@@ -33,8 +33,13 @@ dependencies {
+    androidTestImplementation 'com.android.support.test:rules:1.0.2'
     androidTestImplementation 'com.android.support.test:runner:1.0.2'
     androidTestImplementation 'com.android.support.test.espresso:espresso-core:3.0.2'
+    androidTestImplementation 'org.mockito:mockito-core:2.19.0'
```
👍 Mockito is really good!

arcanerogue commented:
Hi @grasmussen-phunware .  I’ve made some changes based on your recommendations

**Changes:**
* Reworked the getter and added a setter in EventViewModel

* Added error handling to address the case of the id passed to the EventViewModel being invalid or missing. If the Response Body of the network call does not contain an id field, the event will not be inserted into the local database.  When navigating to an Event Detail screen, if an event with the supplied id cannot be found, the user is taken to an "empty" view and a toast message is displayed informing the user that the event was not found.

**Notes:**
State handling of data will be added in a future PR for network and database resources.

**KNOWN ISSUE:** 
If the id field holds a value of text such has "three," the onFailure callback in EventRepository will be called and the exception logged. In the next PR, I will include State Handling to cover this case so that the View knows to display an error message to the user.

grasmussen-phunware commented on app/src/main/java/com/glopez/phunapp/model/EventRepository.kt:
```diff
@@ -46,7 +47,11 @@ class EventRepository(context: Context, eventDatabase: EventDatabase) {
-           AsyncTask.execute{ eventDao.insert(event) }
+            // If the id field was not present in the Response object, the default value of 0 will be set.
+            // If this is the case, the Event will not be inserted into the database.
+            if (event.id > 0) {
```
Hi @arcanerogue I'm glad you are catching this case 👍 

I have two recommendations for you:
1. Turning the hardcoded `0` into  constant and using the constant here instead.  The benefit of doing this kind of change in code bases is helping reduce possible long-term maintenance overhead.
2. Add an else case  with a log statement to log out the info for the event, which has an id that is less than 1.

grasmussen-phunware commented on app/src/main/java/com/glopez/phunapp/view/activities/EventDetailActivity.kt:
```diff
@@ -0,0 +1,129 @@
+    private fun observeEventDetail(eventId: Int) {
+        eventDetailViewModel.getEvent(eventId).observe(this, Observer { event ->
+            event.let {
+                if (it != null && it.id > 0) {
```
@arcanerogue Refer to my recommendation above about the hardcoded `0` as the constant addition applies here also.

grasmussen-phunware commented on app/src/main/java/com/glopez/phunapp/view/activities/EventDetailActivity.kt:
```diff
@@ -0,0 +1,129 @@
+    /**
+     * Displays an "empty" View when the database cannot locate an Event to display to the user
+     */
+    private fun handleViewsOnError() {
```
👍
